RBAC Support for Actions

simster7 opened the pull request:
See https://github.com/argoproj/argo-cd/issues/2002

jessesuen commented on server/rbacpolicy/rbacpolicy.go:
```diff
@@ -25,6 +25,8 @@ const (
 	ActionSync     = "sync"
 	ActionOverride = "override"
+
+	ActionActionType = "action"
```
Unfortunately, this should be named ActionAction for consistency, and if we ever want to use reflection to get these values.

alexec commented:
CI failure looks unrelated to change. Re-building.

evrardjp commented:
Not sure which tests are running, but it seems there is an error in build just after a git diff showing a spacing issue
``
-	ActionAction = "action"
+	ActionAction   = "action"
``

simster7 commented:
Thanks @evrardjp. Seems current failure is due to unrelated tests? @alexec

alexec commented:
The failed test is known flaky.

simster7 commented:
Perfect. @jessesuen should be ready to merge

jessesuen commented on server/application/application.go:
```diff
@@ -1093,7 +1093,8 @@ func (s *Server) RunResourceAction(ctx context.Context, q *application.ResourceA
 	}
-	res, config, _, err := s.getAppResource(ctx, rbacpolicy.ActionOverride, resourceRequest)
+	actionRequest := fmt.Sprintf("%s/%s", rbacpolicy.ActionAction, q.Action)
+	res, config, _, err := s.getAppResource(ctx, actionRequest, resourceRequest)
```
Thinking about this more, `action/<actionname>` is not specific enough. We may have the same action name "pause", which works for both a Deployment, Rollout, and CronJob object. Granting "pause" privileges to all object types is potentially undesirable. We need a convention for actions which reflect the finer grained rbac. For example:

`action/argoproj.io/Rollout/resume`

simster7 replied:
Responded in https://github.com/argoproj/argo-cd/issues/2181

jessesuen replied:
As I understand, the convention being introduced is `action/argoproj.io/Rollout:resume` instead of `action/argoproj.io/Rollout/resume`. I feel the slash is more consistent with all our other RBAC as opposed to the colon.

For example, rbac rules may be:
`action/argoproj.io/Rollout/*`

Which seems better than:
`action/argoproj.io/Rollout:*`

simster7 replied:
Switched convention back to `action/argoproj.io/Rollout/*`

jessesuen commented:
For the UI to understand available actions. We need to implement action discovery better, as described in item (2) in https://github.com/argoproj/argo-cd/issues/2181

simster7 commented:
@jessesuen @alexec Sorry for the crazy delay, but points 1 and 2 from Jesse's issue (#2110) are now done. I intend to open a separate PR for the UI changes. A few points:

1. The new syntax for actions is `group/kind:actionName`. Because of this, I have removed the `group` and `kind` parameters in `argocd app actions run` as they are now a part of the action name. Therefore, **this is currently a breaking change** as the call
```sh
argocd app actions run APPNAME resume --kind Rollout
```
is no longer supported and instead must be
```sh
argocd app actions run APPNAME argoproj.io/Rollout:resume
```
Not sure how to approach this, should I allow backwards compatibility with `resume` commands?

2. `argocd app actions list APPNAME` now lists all actions across all resources by default and shows if they are available:
```sh
$ argocd app actions list blue-green

GROUP        KIND     NAME                       ACTION                      AVAILABLE
argoproj.io  Rollout  blue-green-helm-guestbook  argoproj.io/Rollout:resume  true
```

3. Running an action no longer checks if it is available or not.

jessesuen commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -954,8 +954,9 @@ type ResourceActionDefinition struct {
-	Params []ResourceActionParam `json:"params,omitempty" protobuf:"bytes,2,rep,name=params"`
+	Name      string                `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	Params    []ResourceActionParam `json:"params,omitempty" protobuf:"bytes,2,rep,name=params"`
+	Available bool                  `protobuf:"varint,3,opt,name=available"`
```
This needs to have json struct tags.

jessesuen commented on cmd/argocd/commands/app_actions.go:
```diff
@@ -156,10 +150,23 @@ func NewApplicationResourceActionsRunCommand(clientOpts *argocdclient.ClientOpti
+	if len(groupKindSplit) != 2 {
+		log.Fatal("Action name is malformed")
+	}
+	return groupKindSplit[0], groupKindSplit[1], actionName
```
It appears we are making breaking changes here by requiring a new convention for running actions. I think it's okay if we start introducing a new convention, but we may need to preserve previous flags for backwards compatibility since many pipelines are depending on this behavior.

simster7 replied:
Introduced backwards compatibility with the following message:

```
$ argocd app actions run blue-green resume --kind Rollout

Warning: "resume" action has been deprecated. Please run the action as

	argocd app run blue-green argoproj.io/Rollout/resume

```

jessesuen commented on server/application/application.go:
```diff
@@ -1116,13 +1122,6 @@ func (s *Server) RunResourceAction(ctx context.Context, q *application.ResourceA
-	}
-	if len(filteredAvailableActions) == 0 {
-		return nil, status.Errorf(codes.InvalidArgument, "action not available on resource")
-	}
```
Thanks for removing this check.

jessesuen commented:
My review comments were made without first reading comment https://github.com/argoproj/argo-cd/pull/2110#issuecomment-531953165. But yes, I am glad you recognized the breaking change.

> is no longer supported and instead must be `argocd app actions run APPNAME argoproj.io/Rollout:resume`

How does this work when a user wants to run it against a specific rollout by name?

simster7 commented:
@jessesuen Addressed all of your comments.

> How does this work when a user wants to run it against a specific rollout by name?

There is a `--resource-name` argument available

simster7 commented:
@alexec @jessesuen This should also be good to go!

alexec commented:
@jessesuen do you want to merge please?

alexec commented:
Yay!
